File: java/complex-numbers/src/main/java/ComplexNumber.java
class ComplexNumber {
    private double real;
    private double imag;

    ComplexNumber(double real, double imag) {
        this.real = real;
        this.imag = imag;
    }

    double getReal() {
        return real;
    }

    double getImag() {
        return imag;
    }

    ComplexNumber conjugate() {
        return new ComplexNumber(real, -imag);
    }

    double abs() {
        return Math.sqrt(real * real + imag * imag);
    }

    ComplexNumber add(ComplexNumber other) {
        return new ComplexNumber(real + other.real, imag + other.imag);
    }

    ComplexNumber minus(ComplexNumber other) {
        return new ComplexNumber(real - other.real, imag - other.imag);
    }

    ComplexNumber times(ComplexNumber other) {
        double resReal = real * other.real - imag * other.imag;
        double resImag = imag * other.real + real * other.imag;
        return new ComplexNumber(resReal, resImag);
    }

    ComplexNumber div(ComplexNumber other) {
        double divisor = other.real * other.real + other.imag * other.imag;
        double resReal = (real * other.real + imag * other.imag) / divisor;
        double resImag = (imag * other.real - real * other.imag) / divisor;
        return new ComplexNumber(resReal, resImag);
    }

    ComplexNumber exponentialOf() {
        return new ComplexNumber(Math.exp(real), 0.0).times(new ComplexNumber(Math.cos(imag), Math.sin(imag)));
    }
}

File: kotlin/complex-numbers/src/main/kotlin/ComplexNumber.kt
import kotlin.math.*

data class ComplexNumber(val real: Double = 0.0, val imag: Double = 0.0) {
    val abs: Double
        get() = sqrt(real.pow(2) + imag.pow(2))

    operator fun plus(z: ComplexNumber) = ComplexNumber(real + z.real, imag + z.imag)
    operator fun minus(z: ComplexNumber) = ComplexNumber(real - z.real, imag - z.imag)

    operator fun times(z: ComplexNumber) = ComplexNumber(
        real = (real * z.real) - (imag * z.imag),
        imag = (imag * z.real) + (real * z.imag)
    )
    operator fun div(z: ComplexNumber) = ComplexNumber(
        real = ((real * z.real) + (imag * z.imag)) / (z.real.pow(2) + z.imag.pow(2)),
        imag = ((imag * z.real) - (real * z.imag)) / (z.real.pow(2) + z.imag.pow(2))
    )

    fun conjugate() = ComplexNumber(real, -imag)
}

fun exponential(z: ComplexNumber): ComplexNumber
    = ComplexNumber(real = exp(z.real)) * ComplexNumber(cos(z.imag), sin(z.imag))
